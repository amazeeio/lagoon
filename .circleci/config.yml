version: 2
jobs:
  build:
    machine:
      enabled: true
      image: circleci/classic:edge

    steps:
      - checkout

      - run:
          name: login to dockerhub
          command: docker login -u amazeeiojenkins -p $DOCKERHUB_PASSWORD

      - run:
          name: load cache docker images from dockerhub
          command: make pull -j10 TAG=$CIRCLE_BRANCH

      - run:
          name: show existing images
          command: docker image ls -a --digests

      - run:
          name: check docker
          command: |
            sudo sh -c "echo \"DOCKER_OPTS='--insecure-registry 172.30.0.0/16'\" >> /etc/default/docker"
            sudo service docker restart
            sudo mount --make-rshared /

      - run:
          name: Start openshift
          command: make start-openshift

      - run:
          name: OpenShift Logs
          command: docker logs -f origin
          background: true

      - run:
          name: Build Images
          command: make build -j10 TAG=$CIRCLE_BRANCH

      - run:
          name: push cache docker images to dockerhub
          command: make tag-push -j6 TAG=$CIRCLE_BRANCH
          background: true

      - run:
          name: starting services
          command: make up

      - run:
          name: docker-compose logs
          command: while true; do docker-compose -p lagoon logs -f; done
          background: true

      - run:
          name: give services time to boot up
          command: sleep 20

      - run:
          name: run tests
          command: make tests -j2


