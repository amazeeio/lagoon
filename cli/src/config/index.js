// @flow

import R from 'ramda';
import { readConfigFile } from '../configFile/readConfigFile';
import { globalOptionDefaults } from './globalOptions';

import type { Argv } from 'yargs';
import type { Config } from '../types/Config';

let argv = {};

export const getConfig = (() => {
  let config;

  return ({
    // Dynamic options are options that will likely change every time and should only be specified dynamically.
    // Example: the --overwrite option is allowed to be passed as a command line option, but not allowed to be saved in and read from the config file
    // Example: the --token <path> option is allowed to be passed as a command line option and environment variable, but not allowed to be read from the config file
    dynamicOptionKeys,
  }: {
    dynamicOptionKeys?: Array<string>,
  } = {}): Config => {
    if (!config) {
      if (R.isEmpty(argv)) {
        throw new Error(
          'getConfig() called before argv initialized (setConfig() not yet called?)',
        );
      }
      config = Object.freeze({
        // Manually apply defaults instead of using yargs' `default` option in order to allow config file to override global defaults
        // An alternative would be to move everything into yargs (global options, command options, config using yargs.config, environment variables using yargs.env), but this doesn't allow us an easy way to filter out values from only some sources later with the dynamicOptionsKeys option./
        // Maybe just add an option to filter the configFile to readConfigFile?
        ...globalOptionDefaults,
        ...R.omit(dynamicOptionKeys || [], readConfigFile() || {}),
        ...argv,
      });
    }

    return config;
  };
})();

export function setConfig({
  argv: newArgv,
  dynamicOptionKeys,
}: {
  argv: Argv,
  dynamicOptionKeys: Array<string>,
}): Config {
  // Omit yargs-specific properties
  argv = R.omit(['_', '$0'], newArgv);
  return getConfig({ dynamicOptionKeys });
}
