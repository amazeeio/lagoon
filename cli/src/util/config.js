// @flow

import { fileExists, writeFile } from './fs';
import yaml from 'js-yaml';

import path from 'path';
import co from 'co';

type DeployTask = {
  before_deploy: Array<string>,
  after_deploy: Array<string>,
};

// TODO: Type the rest of the config
export type AmazeeConfig = {
  sitegroup: string,
  deploy_tasks: {
    [name: string]: DeployTask,
  },
};

export type AmazeeConfigInput = {
  sitegroup: string,
};

export function createConfig(
  filepath: string,
  inputOptions: AmazeeConfigInput,
): Promise<void> {
  const config: AmazeeConfig = {
    ...inputOptions,
    deploy_tasks: {
      task1: {
        before_deploy: [],
        after_deploy: [],
      },
    },
  };
  const yamlConfig = yaml.safeDump(config);
  return writeFile(filepath, yamlConfig);
}

function* walker(
  dir: string,
  filename: string,
  root: string,
): Generator<*, ?string, *> {
  let next = dir;

  while (true) {
    const full = path.join(next, filename);
    const exists = yield fileExists(full);

    if (exists) {
      return full;
    } else if (next === root) {
      // ${root}/${filename} does not exist, quit
      return null;
    }

    // Otherwise go up one level
    next = path.dirname(next);
  }
}

/**
 * Tries to resolve a given filename, starting at cwd.
 * Will try ${cwd}/$filename, if not found it will step up
 * one directory (..) until it reaches root (/${filename}).
 *
 * If no file was found, it will return null, otherwise the path
 * of the found file.
 */
export async function findConfig(
  filename: string,
  cwd: string,
): Promise<?string> {
  const start = path.join(cwd, filename);
  const { root } = path.parse(start);

  return co(walker(cwd, filename, root));
}

export function parseConfig(yamlContent: string): AmazeeConfig {
  // TODO: Add schema validation in there if necessary
  return yaml.safeLoad(yamlContent);
}
