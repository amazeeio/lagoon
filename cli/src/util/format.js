// @flow

import R from 'ramda';
import {
  FORMAT,
  FORMAT_CHOICE_TABLE,
  FORMAT_CHOICE_SIMPLE,
  FORMAT_CHOICE_JSON,
  FORMAT_CHOICE_CSV,
} from '../config/globalOptions';
import { getConfig } from '../config';

import type { FormatData } from '../types/Format';

export default function format(data: FormatData) {
  return R.cond([
    [
      R.equals(FORMAT_CHOICE_TABLE),
      () => require('../formatters/table').default(data),
    ],
    [
      R.equals(FORMAT_CHOICE_SIMPLE),
      () => require('../formatters/simple').default(data),
    ],
    [
      R.equals(FORMAT_CHOICE_JSON),
      () => require('../formatters/json').default(data),
    ],
    [
      R.equals(FORMAT_CHOICE_CSV),
      () => require('../formatters/csv').default(data),
    ],
  ])(R.prop(FORMAT, getConfig()));
}
