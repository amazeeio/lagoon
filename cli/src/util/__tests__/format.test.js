// @flow

import format from '../format';
import { getConfig } from '../../config';
import formatAsTable from '../../formatters/table';
import formatAsSimple from '../../formatters/simple';
import formatAsJson from '../../formatters/json';
import formatAsCsv from '../../formatters/csv';

jest.mock('../../config');
jest.mock('../../formatters/table');
jest.mock('../../formatters/simple');
jest.mock('../../formatters/json');
jest.mock('../../formatters/csv');

// Flow does not know which objects are actual mocks
// this function casts given parameter to JestMockFn
const _castMockForFlow = (mockFn: any): JestMockFn<any, any> => mockFn;

const formatAsTableMock = _castMockForFlow(formatAsTable);
const formatAsSimpleMock = _castMockForFlow(formatAsSimple);
const formatAsJsonMock = _castMockForFlow(formatAsJson);
const formatAsCsvMock = _castMockForFlow(formatAsCsv);

describe('format', () => {
  afterEach(() => {
    formatAsTableMock.mockClear();
    formatAsSimpleMock.mockClear();
    formatAsJsonMock.mockClear();
    formatAsCsvMock.mockClear();
  });

  it('should call table formatter with correct configuration', () => {
    _castMockForFlow(getConfig).mockImplementationOnce(() => ({
      format: 'table',
    }));
    format([
      ['Name', 'Customer'],
      ['test-project-table', 'test-customer-table'],
    ]);
    expect(formatAsTableMock.mock.calls).toMatchSnapshot();
    expect(formatAsSimpleMock.mock.calls.length).toBe(0);
    expect(formatAsJsonMock.mock.calls.length).toBe(0);
    expect(formatAsCsvMock.mock.calls.length).toBe(0);
  });

  it('should call simple formatter with correct configuration', () => {
    _castMockForFlow(getConfig).mockImplementationOnce(() => ({
      format: 'simple',
    }));
    format([
      ['Name', 'Customer'],
      ['test-project-simple', 'test-customer-simple'],
    ]);
    expect(formatAsSimpleMock.mock.calls).toMatchSnapshot();
    expect(formatAsTableMock.mock.calls.length).toBe(0);
    expect(formatAsJsonMock.mock.calls.length).toBe(0);
    expect(formatAsCsvMock.mock.calls.length).toBe(0);
  });

  it('should call json formatter with correct configuration', () => {
    _castMockForFlow(getConfig).mockImplementationOnce(() => ({
      format: 'json',
    }));
    format([['Name', 'Customer'], ['test-project-json', 'test-customer-json']]);
    expect(formatAsJsonMock.mock.calls).toMatchSnapshot();
    expect(formatAsTableMock.mock.calls.length).toBe(0);
    expect(formatAsSimpleMock.mock.calls.length).toBe(0);
    expect(formatAsCsvMock.mock.calls.length).toBe(0);
  });

  it('should call csv formatter with correct configuration', () => {
    _castMockForFlow(getConfig).mockImplementationOnce(() => ({
      format: 'csv',
    }));
    format([['Name', 'Customer'], ['test-project-csv', 'test-customer-csv']]);
    expect(formatAsCsvMock.mock.calls).toMatchSnapshot();
    expect(formatAsTableMock.mock.calls.length).toBe(0);
    expect(formatAsSimpleMock.mock.calls.length).toBe(0);
    expect(formatAsJsonMock.mock.calls.length).toBe(0);
  });

  it("shouldn't call any formatter with an incorrect configuration", () => {
    _castMockForFlow(getConfig).mockImplementationOnce(() => ({
      format: 'non-existent formatting option',
    }));
    format([['Name', 'Customer'], ['test-project-none', 'test-customer-none']]);
    expect(formatAsTableMock.mock.calls.length).toBe(0);
    expect(formatAsSimpleMock.mock.calls.length).toBe(0);
    expect(formatAsJsonMock.mock.calls.length).toBe(0);
    expect(formatAsCsvMock.mock.calls.length).toBe(0);
  });
});
