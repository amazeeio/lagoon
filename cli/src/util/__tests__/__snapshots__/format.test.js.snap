// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`format should call csv formatter with correct configuration 1`] = `
Array [
  Array [
    Array [
      Array [
        "Name",
        "Customer",
      ],
      Array [
        "test-project-csv",
        "test-customer-csv",
      ],
    ],
  ],
]
`;

exports[`format should call json formatter with correct configuration 1`] = `
Array [
  Array [
    Array [
      Array [
        "Name",
        "Customer",
      ],
      Array [
        "test-project-json",
        "test-customer-json",
      ],
    ],
  ],
]
`;

exports[`format should call simple formatter with correct configuration 1`] = `
Array [
  Array [
    Array [
      Array [
        "Name",
        "Customer",
      ],
      Array [
        "test-project-simple",
        "test-customer-simple",
      ],
    ],
  ],
]
`;

exports[`format should call table formatter with correct configuration 1`] = `
Array [
  Array [
    Array [
      Array [
        "Name",
        "Customer",
      ],
      Array [
        "test-project-table",
        "test-customer-table",
      ],
    ],
  ],
]
`;
